src/StreamJsonRpc/WebSocketMessageHandler.cs

```diff
@@ -0,0 +1,137 @@
+                // The message exceeds the size of our buffer.
+                if (this.readDecoder == null)
+                {
+                    this.readDecoder = this.Encoding.GetDecoder();
+                }
+
+                var jsonBuilder = new StringBuilder();
+                char[] decodedChars = new char[this.readBuffer.Array.Length];
+                void DecodeInput()
+                {
+                    int decodedCharsCount = this.readDecoder.GetChars(this.readBuffer.Array, 0, result.Count, decodedChars, 0, result.EndOfMessage);
+                    jsonBuilder.Append(decodedChars, 0, decodedCharsCount);
+                }
+
+                DecodeInput();
+                while (!result.EndOfMessage)
+                {
+                    result = await this.WebSocket.ReceiveAsync(this.readBuffer, cancellationToken).ConfigureAwait(false);
+                    DecodeInput();
+                }
+
+                return jsonBuilder.ToString();
+            }
+        }
+
+        /// <inheritdoc />
+        protected async override Task WriteCoreAsync(string content, Encoding contentEncoding, CancellationToken cancellationToken)
+        {
+            Requires.NotNull(content, nameof(content));
+            Requires.NotNull(contentEncoding, nameof(contentEncoding));
+
+            if (contentEncoding.GetByteCount(content) <= this.writeBuffer.Length)
+            {
+                // Fast path: send the whole message as a single chunk.
+                int bytesWritten = contentEncoding.GetBytes(content, 0, content.Length, this.writeBuffer, 0);
+                var bufferSegment = new ArraySegment<byte>(this.writeBuffer, 0, bytesWritten);
+                await this.WebSocket.SendAsync(bufferSegment, WebSocketMessageType.Text, true, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                var encoder = contentEncoding.GetEncoder();
+                bool completed;
+                int bytesUsed;
+                int charsLeftToConvert = content.Length;
+                while (charsLeftToConvert > 0)
+                {
+                    unsafe
+                    {
+                        fixed (byte* writeBuffer = this.writeBuffer)
+                        fixed (char* pContent = content)
+                        {
+                            char* pStart = pContent + content.Length - charsLeftToConvert;
+                            encoder.Convert(pStart, charsLeftToConvert, writeBuffer, this.writeBuffer.Length, false, out int charsUsed, out bytesUsed, out completed);
+                            charsLeftToConvert -= charsUsed;
+                        }
+                    }
+
+                    var bufferSegment = new ArraySegment<byte>(this.writeBuffer, 0, bytesUsed);
+                    await this.WebSocket.SendAsync(bufferSegment, WebSocketMessageType.Text, completed, cancellationToken).ConfigureAwait(false);
+                }
+            }
```

Comment:
huh, I didn't have to do this.  WebSocket seems to send whatever size message I give it...I did up to 50k sized messages without this logic..?
        {
            var buffer = Encoding.UTF8.GetBytes(content);
            await this.WebSocket.SendAsync(new ArraySegment<byte>(buffer), WebSocketMessageType.Text, true, cancellationToken)
                .ConfigureAwait(false);
        }

Reply:
Yup, yours is simpler. But it requires allocating a new buffer for each and every transmitted message. My approach reuses a buffer for each message. That reduces GC pressure and thus should promote high throughput when you experience memory pressure.

Reply:
> unsafe buffers make me nervous
Well, that's why it's useful to have it written in a well-tested library so you can depend on it rather than own it yourself. :) Everything you do in .NET ultimately becomes "unsafe" at some level. In this case, I have to use unsafe because the encoder doesn't offer an overload that takes `byte[]` and an offset, so the only way to get buffer offset behavior is to use pointers. 

> why not increase the write buffer to size and then do simple write

Reliable perf and high throughput.